From 4d3e04bb158063e156c6ff6dbb29d3985f5a26f6 Mon Sep 17 00:00:00 2001
From: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Date: Mon, 10 Sep 2018 16:08:37 +0300
Subject: [PATCH 4683/5242] MLK-18655-5 sound: asoc: add hwvad noise
 decimation amixer control

commit  bb24f099aaba679b78802e50a8ef141e4b41529a from
https://source.codeaurora.org/external/imx/linux-imx.git

Add control for Voice Activity Detector Noise OR Enable.
Enables a OR logic in the output of minimum noise estimator block. For
more information see Filter Result Gain Setting on page 71 - micfil
reference manual:
0b - Noise input is not decimated.
1b - Noise input is decimated.

Signed-off-by: Cosmin-Gabriel Samoila <cosmin.samoila@nxp.com>
Reviewed-by: Daniel Baluta <daniel.baluta@nxp.com>
(cherry picked from commit 317d2226fcfbbd8ae91eac9f685d71663227ae2a)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 sound/soc/fsl/fsl_micfil.c |   11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/sound/soc/fsl/fsl_micfil.c b/sound/soc/fsl/fsl_micfil.c
index e25513f..9c6a4ae 100644
--- a/sound/soc/fsl/fsl_micfil.c
+++ b/sound/soc/fsl/fsl_micfil.c
@@ -121,6 +121,10 @@ struct fsl_micfil_soc_data {
 	"OFF", "ON",
 };
 
+static const char * const micfil_hwvad_noise_decimation[] = {
+	"Disabled", "Enabled",
+};
+
 
 static const struct soc_enum fsl_micfil_enum[] = {
 	SOC_ENUM_SINGLE(REG_MICFIL_CTRL2,
@@ -135,6 +139,10 @@ struct fsl_micfil_soc_data {
 			    micfil_hwvad_zcd_enable),
 	SOC_ENUM_SINGLE_EXT(ARRAY_SIZE(micfil_hwvad_zcdauto_enable),
 			    micfil_hwvad_zcd_enable),
+	SOC_ENUM_SINGLE(REG_MICFIL_VAD0_NCONFIG,
+			MICFIL_VAD0_NCONFIG_NOREN_SHIFT,
+			ARRAY_SIZE(micfil_hwvad_noise_decimation),
+			micfil_hwvad_noise_decimation),
 };
 
 static int set_quality(struct snd_kcontrol *kcontrol,
@@ -549,7 +557,8 @@ static int hwvad_get_zcd_adj(struct snd_kcontrol *kcontrol,
 		     hwvad_get_zcd_en, hwvad_put_zcd_en),
 	SOC_ENUM_EXT("HWVAD Zero-Crossing Detector Auto Threshold", fsl_micfil_enum[4],
 		     hwvad_get_zcd_auto, hwvad_put_zcd_auto),
-
+	SOC_ENUM_EXT("HWVAD Noise OR Enable", fsl_micfil_enum[5],
+		     snd_soc_get_enum_double, snd_soc_put_enum_double),
 	{
 		.iface = SNDRV_CTL_ELEM_IFACE_MIXER,
 		.name = "HWVAD Input Gain",
-- 
1.7.9.5

